### PR TITLE
fix(automodel): avoid fp32 weight load when optimizer holds master

### LINKED ISSUE #2865: [Automodel]: Hard-coded fp32 model weights doubles memory with FusedAdam
## Summary

The automodel backend hard-codes the model load dtype to **fp32** in order to support master weights [here](https://github.com/NVIDIA-NeMo/RL/blob/1ff6e117a41e37fdd80401ae375ec9f23e8d28ea/nemo_rl/models/automodel/setup.py#L305). 
However, the optimizer is configurable and supports both `torch.optim.AdamW` and TE `FusedAdam`. When **FusedAdam** is used, the fp32 load is unnecessary and causes **~2× optimizer master-weight memory** (and ~2× model-weight memory), because FusedAdam already maintains its own fp32 master internally.

## Details

With the TE FusedAdam optimizer, the fp32 master is kept **inside the optimizer**, via:

```yaml
optimizer:
  name: transformer_engine.pytorch.optimizers.fused_adam.FusedAdam
  kwargs:
    master_weights: true
    store_param_remainders: true
```

`store_param_remainders: true` lets FusedAdam split the fp32 master into **bf16 (high 16 bits) + int16 remainder (low 16 bits)** — `16 + 16 = 32 = fp32`. This depends on the parameters being **bf16**: once a parameter is fp32, TE detects `param.dtype != bf16` and **skips the remainder path, storing a separate full fp32 master** instead. Ref: [https://github.com/NVIDIA/TransformerEngine/blob/42b840051647eef89761a16dfdff87e82bb253ab/transformer_engine/pytorch/optimizers/fused_adam.py#L97](https://github.com/NVIDIA/TransformerEngine/blob/42b840051647eef89761a16dfdff87e82bb253ab/transformer_engine/pytorch/optimizers/fused_adam.py#L97)

So when the load dtype is hard-coded to fp32, with FusedAdam:

- **Model weights**: bf16 → fp32 = **2× memory**.
- **Optimizer master**: `bf16 + int16 remainder` → **full fp32** = **2× memory**.

## Reproduction

Run a grpo training, toggling only the model load dtype (`AutoConfig(torch_dtype=...)` in `nemo_rl/models/automodel/setup.py`) between
`float32` and `bfloat16`, and sum the `optimizer.state` tensor bytes after the first `optimizer.step()`.

## Observed (rank 0)

| metric                    | fp32 load (current)              | modified to bf16 load |
| ------------------------- | -------------------------------- | -------------------- |
| `master_param`            | **fp32 7.32 GB** + int16 0.60 GB | int16 — 4.26 GB      |
| optimizer state /rank     | **16.43 GB**                     | 12.77 GB             |
| optimizer state all-ranks | 131.5 GB                         | 102.2 GB             |
| model params all-ranks    | 66.6 GB                          | 35.7 GB              |

`exp_avg` / `exp_avg_sq` are unchanged (bf16); the entire increase is the master copy.

## Suggested fix

Choose the model load dtype based on the optimizer (AdamW vs FusedAdam) instead of hard-coding fp32: load fp32 only for optimizers without an internal master, and keep the compute dtype (bf16) when FusedAdam holds the master.

For reference, I have a candidate fix in: [https://github.com/NVIDIA-NeMo/RL/commit/e968d90f72925637e8834f81510b5f7a181b4093](https://github.com/NVIDIA-NeMo/RL/commit/e968d90f72925637e8834f81510b5f7a181b4093)

### PR DESCRIPTION
**Pull request by lonexreb**
## What
The automodel (DTensor) backend hard-coded the model load dtype to `float32` in `nemo_rl/models/automodel/setup.py` so the optimizer could keep fp32 master weights. When the optimizer maintains its **own** fp32 master — TE `FusedAdam` with `master_weights=True` (bf16 params + int16 remainder) — that fp32 load is redundant and roughly **doubles** both model-weight and optimizer master-weight memory.

Fixes #2865.

## Change
Pick the load dtype from the optimizer config in `validate_and_prepare_config`:
- `master_weights=True` (optimizer holds the master) → load in the **compute dtype**.
- plain AdamW / no `master_weights` / no optimizer configured → keep the **fp32** load (unchanged behavior).

Only the load dtype changes; the FSDP mixed-precision policy and everything downstream are untouched.

## Impact
No behavior change for the exemplar configs (they use `torch.optim.AdamW`, which has no internal master → still fp32). The savings apply only to the FusedAdam+`master_weights` path the issue reports (~131GB → ~102GB optimizer state, ~67GB → ~36GB model params in the reporter's run).

## Test plan
- [x] New unit tests in `tests/unit/models/automodel/test_automodel_setup.py` asserting the `torch_dtype` passed to `AutoConfig.from_pretrained`:
  - plain AdamW → `float32`
  - FusedAdam + `master_weights=True` → compute dtype (bf16)
  - optimizer absent → `float32`
- [ ] CI (`/ok to test`)

_Note: local test run blocked by an unrelated `uv`/Python `mac_ver()` issue on the dev box; logic validated in isolation and syntax-checked. CI will exercise the full suite._

### CODE REVIEW

**Review by terrykong — COMMENTED** [review]
@hemildesai could you review?

**Comment by lonexreb** [comment]
Gentle ping @hemildesai @terrykong — this has been green and mergeable since mid-July (all checks passing, no conflicts). Happy to address any feedback or rebase if you'd prefer it freshened before review.

**Comment by lonexreb** [comment]
@hemildesai @yuki-97 — gentle ping when you have bandwidth. This fixes #2865: the automodel path loads weights in fp32 even when the precision-aware optimizer already holds fp32 master weights, doubling load-time memory for no benefit. All pre-CI checks are green and the branch is mergeable against current main; it still needs a `CI:*` label to run the full tier. Happy to address any feedback or rebase again if it goes stale.
